recaf-assembler/src/main/java/me/coley/recaf/assemble/transformer/JasmToAstTransformer.java

```diff
@@ -0,0 +1,415 @@
+        add(new LineInstruction(-1, label.content(), line.asInt()));
+    }
+
+    @Override
+    public void visitMultiANewArrayInsn(String desc, int dims) throws AssemblerException {
+        add(new MultiArrayInstruction(Opcodes.MULTIANEWARRAY, desc, dims));
+    }
+
+    public HandleInfo from(HandleGroup handle) throws AssemblerException {
+        if(handle.isField()) {
+            FieldDescriptor fs = new FieldDescriptor(content(handle.getDescriptor()));
+            return new HandleInfo(handle.getHandleType().content(), fs.owner, fs.name, content(handle.getFieldDescriptor()));
+        }else {
+            MethodDescriptor mdh = new MethodDescriptor(handle.getDescriptor().content(), false);
+            return new HandleInfo(
+                    handle.getHandleType().content(),
+                    mdh.owner,
+                    mdh.name,
+                    mdh.getDescriptor());
+        }
+    }
+
+    @Override
+    public void visitInvokeDyanmicInsn(String identifier, IdentifierGroup descriptor, HandleGroup handle, ArgsGroup args) throws AssemblerException {
+        HandleInfo handleInfo = from(handle);
+        List<IndyInstruction.BsmArg> bsmArgs = new ArrayList<>();
+        for (Group arg : args.getBody().getChildren()) {
+            bsmArgs.add(new IndyInstruction.BsmArg(from(arg), convert(arg)));
+        }
+        add(new IndyInstruction(
+                Opcodes.INVOKEDYNAMIC,
+                identifier,
+                descriptor.content(),
+                handleInfo,
+                bsmArgs));
+    }
+
+    public ArgType from(Group group) {
+        if(group instanceof NumberGroup) {
+            NumberGroup number = (NumberGroup) group;
+            if(number.isFloat()) {
+                return number.isWide() ? ArgType.DOUBLE : ArgType.FLOAT;
+            }else {
+                return number.isWide() ? ArgType.LONG : ArgType.INTEGER;
+            }
+        }else if(group instanceof StringGroup) {
+            return ArgType.STRING;
+        }else if(group instanceof TypeGroup) {
+            return ArgType.TYPE;
+        }else if(group instanceof HandleGroup) {
+            return ArgType.HANDLE;
+        }
+        throw new IllegalArgumentException("Cannot convert to constant '" + group.content() + "'");
+    }
+
+    @Override
+    public void visitInsn(int opcode) throws AssemblerException {
+        add(new Instruction(opcode));
+    }
+
+    @Override
+    public void visitExpr(ExprGroup expr) throws AssemblerException {
+        add(new Expression(expr.textGroup.content()));
+    }
+
+    @Override
+    public void visitClass(AccessModsGroup accessMods, IdentifierGroup identifier) throws AssemblerException {
+
+    }
+
+    @Override
+    public void visitSuper(ExtendsGroup extendsGroup) throws AssemblerException {
+
+    }
+
+    @Override
+    public void visitImplements(ImplementsGroup implementsGroup) throws AssemblerException {
+
```

Comment:
Are these supposed to be empty?

Reply:
The plan is to get field/method assembly in, and then come back and do class level assembly after.
The base of class level assembly is set up but not fully implemented as seen here and in a few other places.